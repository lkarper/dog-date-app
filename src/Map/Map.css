.Map__map-container {
    width: 500px;
    height: 500px;
}

.Map__map-outer-container {
    height: 500px;
    width: 500px;
    margin: 0;
    position: relative;;
}

.Map__marker {
    display: block;
    border: none;
    cursor: pointer;
    padding: 0;
    background-image: url('/images/map-marker.png');
    background-size: cover;
    width: 50px;
    height: 50px;
}

.Map__coordinates {
    background: rgba(0, 0, 0, 0.7);
    color: #fff;
    position: absolute;
    bottom: 0px;
    padding: 5px 10px;
    margin: 0;
    font-size: 14px;
    line-height: 18px;
    border-radius: 3px;
}