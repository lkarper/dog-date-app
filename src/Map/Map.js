import React, { useRef, useEffect, useState } from 'react';
import mapboxgl from 'mapbox-gl';
import ForwardGeocodeAddress from '../ForwardGeocodeAddress/ForwardGeocodeAddress';
import config from '../config';
import './Map.css';

const Map = (props) => {

    const { tempCoordinates, setTempCoordinates } = props;

    const [map, setMap] = useState(null);
    const [marker, setMarker] = useState(null);
    const [markerCoordinates, setMarkerCoordinates] = useState({ lat: 40.7812, lon: -73.9665 });

    let mapContainer = useRef(null);
    let markerRef = useRef(null);

    const setRefMap = (el) => {
        mapContainer = el;
    }

    const setRefMark = (el) => {
        markerRef = el;
    }

    const setMarkerToMyLocation = () => {
        navigator.geolocation.getCurrentPosition((position) => {

            setMarkerCoordinates({
                lat: position.coords.latitude,
                lon: position.coords.longitude 
            });
        });
    }

    useEffect(() => {

        const locationiqKey = config.mapbox_key;

        const map = new mapboxgl.Map({
            container: mapContainer,
            attributionControl: false,
            style: `https://tiles.locationiq.com/v2/streets/vector.json?key=${locationiqKey}`,
            center: [markerCoordinates.lon, markerCoordinates.lat],
            zoom: 12
        });

        setMap(map);

        const marker = new mapboxgl.Marker(markerRef, {
            draggable: true,
        })
            .setLngLat([markerCoordinates.lon, markerCoordinates.lat])
            .addTo(map);
        
        setMarker(marker);

        const onDragEnd = () => {
            const lngLat = marker.getLngLat();
            setMarkerCoordinates({ lat: lngLat.lat, lon: lngLat.lng });
        }

        marker.on('dragend', onDragEnd);

    }, [props.setTempCoordinates, markerCoordinates.lat, markerCoordinates.lon]);

    useEffect(() => {
        if (marker) {
            map.setCenter([markerCoordinates.lon, markerCoordinates.lat]);
            marker.setLngLat([markerCoordinates.lon, markerCoordinates.lat]).addTo(map);
            setMarker(marker);

            const onDragEnd = () => {
                const lngLat = marker.getLngLat();
                setMarkerCoordinates({ lat: lngLat.lat, lon: lngLat.lng });
            }

            marker.on('dragend', onDragEnd);

            if (tempCoordinates.lat !== markerCoordinates.lat && tempCoordinates.lon !== markerCoordinates.lon) {
                const { lat, lon } = markerCoordinates;
                setTempCoordinates({
                    lon,
                    lat
                });
            }
        }
    }, [markerCoordinates, tempCoordinates, setTempCoordinates, map, marker])


    return (
        <div className='Map__search-container'>
            <div className='Map__map-outer-container'>
                <ForwardGeocodeAddress setMarkerCoordinates={setMarkerCoordinates} />
                <div id='map-container' ref={el => setRefMap(el)} className='Map__map-container'>
                </div>
                <div className='Map__marker' ref={el => setRefMark(el)}></div>
                <div id='coordinates' className='Map__coordinates'>Latitude: {markerCoordinates.lat}, Longitude: {markerCoordinates.lon}</div>
            </div>
            <button 
                className='Map__set-marker-button'
                type='button' 
                onClick={setMarkerToMyLocation}
            >
                Set marker to my location
            </button>
        </div>
    );
}

export default Map;