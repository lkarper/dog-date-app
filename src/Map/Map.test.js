import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Map from './Map';

describe('Map component', () => {
    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Map />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = mount(<Map />);
        expect(toJSON(wrapper)).toMatchSnapshot();
        wrapper.unmount();
    });
});
