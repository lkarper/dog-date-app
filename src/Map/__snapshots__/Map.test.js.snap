// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component renders the UI as expected 1`] = `
<Map
  setTempCoordinates={[Function]}
  tempCoordinates={
    Object {
      "lat": 0,
      "lon": 0,
    }
  }
>
  <div
    className="Map__search-container"
  >
    <div
      className="Map__map-outer-container"
    >
      <ForwardGeocodeAddress
        setMarkerCoordinates={[Function]}
      >
        <div
          aria-live="polite"
          className="ForwardGeocodeAddress__outer-container"
        >
          <input
            aria-describedby="current-set-coordinates"
            aria-label="Search for a place by name or address."
            className="ForwardGeocodeAddress__search-input"
            id="place-search"
            name="place-search"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Search by name or address"
            type="text"
            value=""
          />
          <NativeClickListener
            className="ForwardGeocodeAddress__click-container hidden"
            id="ForwardGeocodeAddress__click-container"
            onClick={[Function]}
          >
            <div
              className="ForwardGeocodeAddress__click-container hidden"
              id="ForwardGeocodeAddress__click-container"
              role="alert"
            >
              <ol
                className="ForwardGeocodeAddress__results hidden"
              />
            </div>
          </NativeClickListener>
          <div
            role="alert"
          />
        </div>
      </ForwardGeocodeAddress>
      <div
        className="Map__map-container"
        id="map-container"
      />
      <div
        className="Map__marker"
      />
      <div
        className="Map__coordinates"
        id="coordinates"
      >
        Latitude: 
        0
        , Longitude: 
        0
      </div>
    </div>
    <button
      className="Map__set-marker-button button"
      onClick={[Function]}
      type="button"
    >
      Set marker to my location
    </button>
    <p
      className="Map__marker-draggable-p"
    >
      (Marker is draggable.)
    </p>
  </div>
</Map>
`;
